proposals/p1885.md

```diff
@@ -0,0 +1,458 @@
+                done: i < 0,
+            }),
+        };
+    },
+});
+```
+
+#### Go
+
+Go has non-standard support for range-based with custom types only.
+
+https://stackoverflow.com/questions/35810674/in-go-is-it-possible-to-iterate-over-a-custom-type
+
+## Proposal
+
+Provide interfaces that can be implemented by user types to enable support for
+ranged-for loops.
+
+We propose three interfaces:
+
+-   [**Indexable**](#indexable): Implements index-based random access
+-   [**Iterable**](#iterable): Implements iterator-based sequential access
+-   [**Consumable**](#consumable): Implements single-use sequential access
+
+Users would benefit from providing an `external impl`, to limit class members
+pollution.
+
+## Details
+
+This proposal covers support for 2 different type of containers:
+
+-   Random access:
+    -   `Indexable`
+-   Sequential access:
+    -   `Iterable`, `Iterator`
+    -   `Consumable`
+
+Due to the semantic and functional difference between these interfaces, they
+should not conflict with the [One way](/docs/project/principles/one_way.md).
+
+In absence of strong requirements from the algorithm, this would require future
+algorithms to:
+
+-   A. explicitly support multiple interfaces: expressive, and specialize
+    implementations based on type features (likely preferred)
+-   or B. define adapters or implicit conversions to respectively simplify
+    algorithms implementation or end-user usage
+
+### Indexable
+
+`Indexable` represents a container that supports index-based random access.
+
+#### Range-based `for` behavior
+
+-   Get container size (`size`)
+-   Iterate from `[0;size[`, returning current item at each iteration
+
+#### Interface and usage
+
+```rust
+// Carbon definition
+interface Indexable() {
+    let ElementType:! Type
+    fn GetSize[me: Self]() -> u32;
+    fn GetAt[me: Self](pos: u32) -> ElementType;
+}
+
+// Usage
+class MyArray {
+    ...
+    impl as Indexable where .ElementType = i32 {
+        fn GetSize[me: Self]() -> u32 { ... };
+        fn GetAt[me: Self](pos: u32) -> ElementType { ... };
+    }
+}
+
+var array: MyArray = {...};
+for (var element: auto in array) {
+    ...
+}
+```
+
+#### Implicit `[]` support
+
+A related added value would be the support for index-based access and assignment
+with the `[]` operator, as a 1D array, for `Indexable` types. In addition to
+basic single-element operations, this opens the door for future more advance
+uses, as proposed in
+[PR-1928](https://github.com/carbon-language/carbon-lang/pull/1928).
+
+The type-checker would however not be able to validate `Indexable`-based array,
+in comparison to fixed arrays:
+
+```rust
+fn Foo[T:! Indexable](array: T) {
+    array[1] = array[2]; // May or may not exist
+}
+```
+
+#### Mutable access
+
+TODO
+
+### Consumable
+
+An interface representing single-use sequential access type of container. Items
+ca be read (consumed) once, and are not expected to be accessible anymore after
+usage.
```

Comment:
This is a bit different yes, and steps into data ownership territory. The two use cases I had in mind were generators and producer/consumer scenarios, and having an interface to make this explicit.
In terms of common types, this would apply for sure to queues and stacks, as well as any iterable (in the general sense). Generators too, though the semantic added value seems a bit more limited.

From a user standpoint, this could be something like:
```
for (e: auto in queue as Consumable) {
  ...
}
```
Which conveys the intent, independent of the container. In this example I would expect it to just `pop` until empty.
It is definitely a bit more unusual, so I'm not opposed to moving it to a different proposal, changing or dropping it.

Reply:
Parked `Consumable` in 00ef294